packages/contracts-periphery/contracts/universal/op-nft/OptimistInviter.sol

```diff
@@ -0,0 +1,228 @@
+        require(
+            commitments[keccak256(abi.encode(_claimer, _signature))],
+            "OptimistInviter: claimer and signature have not been committed yet"
+        );
+
+        // Generate a EIP712 typed data hash to compare against the signature.
+        bytes32 digest = _hashTypedDataV4(
+            keccak256(
+                abi.encode(
+                    CLAIMABLE_INVITE_TYPEHASH,
+                    _claimableInvite.issuer,
+                    _claimableInvite.nonce
+                )
+            )
+        );
+
+        // Uses SignatureChecker, which supports both regular ECDSA signatures from EOAs as well as
+        // ERC-1271 signatures from contract wallets or multi-sigs. This means that if the issuer
+        // wants to revoke a signature, they can use a smart contract wallet to issue the signature,
+        // then invalidate the signature after issuing it.
+        require(
+            SignatureCheckerUpgradeable.isValidSignatureNow(
+                _claimableInvite.issuer,
+                digest,
+                _signature
+            ),
+            "OptimistInviter: invalid signature"
+        );
+
+        // The issuer includes a pseudorandom nonce in the signature to prevent replay attacks.
+        // This checks that the nonce has not been used for this issuer before. The nonces are
+        // scoped to the issuer address, so the same nonce can be used by different issuers without
+        // clashing.
+        require(
+            !usedNonces[_claimableInvite.issuer][_claimableInvite.nonce],
+            "OptimistInviter: nonce has already been used"
+        );
+
+        // Set the nonce as used for the issuer so that it cannot be replayed.
+        usedNonces[_claimableInvite.issuer][_claimableInvite.nonce] = true;
+
+        // Check the AttestationStation contract to see how many invites the issuer has left.
+        bytes memory attestation = ATTESTATION_STATION.attestations(
+            address(this),
+            _claimableInvite.issuer,
+            bytes32("optimist.can-invite")
+        );
+        // Failing this check means that the issuer was never granted any invites to begin with.
+        require(attestation.length > 0, "OptimistInviter: issuer has no invites");
+
+        uint256 count = abi.decode(attestation, (uint256));
+
+        // Failing this check means that the issuer has used up all of their existing invites.
+        require(count > 0, "OptimistInviter: issuer has no invites");
+
+        // Create the attestation that the claimer can mint from the issuer's invite.
+        // The invite issuer is included in the data of the attestation.
+        ATTESTATION_STATION.attest(
+            _claimer,
+            bytes32("optimist.can-mint-from-invite"),
+            abi.encode(_claimableInvite.issuer)
+        );
+
+        // Reduce the issuer's invite count by 1 by re-attesting the optimist.can-invite attestation
+        // with the new count.
+        count--;
```

Comment:
This  could technically be wrapped in an unchecked block because of the count check above